runtime/integration-tests/src/generic/cases/liquidity_pools.rs

```diff
@@ -1780,23 +1924,293 @@ mod centrifuge {
+					after_transfer_alice,
+					pre_transfer_alice - usdc(TRANSFER_AMOUNT)
+				);
+				assert_eq!(after_transfer_bob, pre_transfer_bob + usdc(TRANSFER_AMOUNT));
+				assert_eq!(after_transfer_charlie, pre_transfer_charlie);
+			});
 		}
 
-		fn restrict_usdc_xcm_transfer<T: Runtime>() {
-			todo!()
+		fn restrict_usdc_xcm_transfer<T: Runtime + FudgeSupport>() {
+			let mut env = FudgeEnv::<T>::from_storage(
+				<paras::GenesisConfig as GenesisBuild<FudgeRelayRuntime<T>>>::build_storage(
+					&paras::GenesisConfig {
+						paras: vec![(
+							1000.into(),
+							ParaGenesisArgs {
+								genesis_head: Default::default(),
+								validation_code: ValidationCode::from(vec![0, 1, 2, 3]),
+								para_kind: ParaKind::Parachain,
+							},
+						)],
+					},
+				)
+				.unwrap(),
+				Genesis::default()
+					.add(genesis::balances::<T>(cfg(10)))
+					.storage(),
+				Default::default(),
+			);
+
+			setup_xcm(&mut env);
+
+			setup_usdc_xcm(&mut env);
+
+			env.sibling_state_mut(|| {
+				register_usdc::<T>();
+			});
+
+			env.parachain_state_mut(|| {
+				register_usdc::<T>();
+
+				let alice_initial_usdc = usdc(3_000);
+
+				assert_ok!(orml_tokens::Pallet::<T>::mint_into(
+					USDC,
+					&Keyring::Alice.into(),
+					alice_initial_usdc
+				));
+
+				assert_ok!(
+					pallet_transfer_allowlist::Pallet::<T>::add_transfer_allowance(
+						RawOrigin::Signed(Keyring::Alice.into()).into(),
+						USDC,
+						Location::XCM(BlakeTwo256::hash(
+							&MultiLocation::new(
+								1,
+								X2(
+									Parachain(T::FudgeHandle::SIBLING_ID),
+									Junction::AccountId32 {
+										id: Keyring::Alice.into(),
+										network: None,
+									}
+								)
+							)
+							.encode()
+						))
+					)
+				);
+
+				assert_noop!(
+					pallet_restricted_xtokens::Pallet::<T>::transfer(
+						RawOrigin::Signed(Keyring::Alice.into()).into(),
+						USDC,
+						usdc(1_000),
+						Box::new(
+							MultiLocation::new(
+								1,
+								X2(
+									Parachain(T::FudgeHandle::SIBLING_ID),
+									Junction::AccountId32 {
+										id: Keyring::Bob.into(),
+										network: None,
+									}
+								)
+							)
+							.into()
+						),
+						WeightLimit::Unlimited,
+					),
+					pallet_transfer_allowlist::Error::<T>::NoAllowanceForDestination
+				);
+
+				assert_ok!(pallet_restricted_xtokens::Pallet::<T>::transfer(
+					RawOrigin::Signed(Keyring::Alice.into()).into(),
+					USDC,
+					usdc(1_000),
+					Box::new(
+						MultiLocation::new(
+							1,
+							X2(
+								Parachain(T::FudgeHandle::SIBLING_ID),
+								Junction::AccountId32 {
+									id: Keyring::Alice.into(),
+									network: None,
+								}
+							)
+						)
+						.into()
+					),
+					WeightLimit::Unlimited,
+				));
+
+				assert_eq!(
+					orml_tokens::Pallet::<T>::free_balance(USDC, &Keyring::Alice.into()),
+					alice_initial_usdc - usdc(1_000),
+				);
+			});
+
+			// NOTE - we cannot confirm that the Alice account present on the
```

Comment:
When executing the above transfer, the xtokens pallet uses the `ToNonReserve` transfer kind. 

Given this transfer kind, xtokens generates an XCM message that has these instructions - `WithdrawAsset`, `ClearOrigin`, `BuyExecution`, `DepositReserveAsset` and sends it to `Parachain(1000)` which is the location of the USDC currency used in this test.

The `DepositReserveAsset` instruction is supposed to then trigger a transfer from Parachain(1000) to our sibling chain.

/cc @mustermeiszer @NunoAlexandre

Reply:
For context:

When a token XT is transferred from Parachain X to parachain Y, X keeps a record that XT went out to Y. Y can not send XT to X that did not leave X to Y. So this behaviour from xtokens makes sense, as otherwise, say, we would send 1) XT from X to Y, then 2) from Y to Z and finally 3) from Z to X; On step 3, X wouldn't know that Z owned XT so it can not accept that incoming transfer.

So the hop is necessary due to the current design of XCM. The goal of this test is to ensure that restricted_xtokens handles the permissions properly for USDC, and all that means is that it passes the request through to XTokens and thereafter outbounds the parachain; Meaning than an OK on the sending parachain is all we need to verify this property; we don't need to ensure that the intended destination chain/account receives the tokens.

Reply:
Yes, that should be fine!